src/BreakpointViewer.cpp

```diff
@@ -438,278 +474,365 @@ void BreakpointViewer::setBreakState()
 	runState = false;
 }
 
-void BreakpointViewer::sync()
+void BreakpointViewer::stretchTable(BreakpointData::Type type)
 {
-	auto sa = ScopedAssign(userMode, false);
-	keepUnusedItems();
-
-	int bpRow = bpTableWidget->rowCount();
-	int wpRow = wpTableWidget->rowCount();
-	int cnRow = cnTableWidget->rowCount();
+	if (type == BreakpointData::ALL) {
+		for (auto* table : selector) {
+			// will not stretch without sorting
+			table->setSortingEnabled(true);
+			table->resizeColumnsToContents();
+		}
+		return;
+	}
 
-	if (breakpoints->breakpointCount() != -1) {
-		for (int index = 0; index < breakpoints->breakpointCount(); ++index) {
-			const Breakpoint& bp = breakpoints->getBreakpoint(index);
+	auto* table = selector[type];
+	// will not stretch without sorting
+	table->setSortingEnabled(true);
+	table->resizeColumnsToContents();
+}
 
-			switch (bp.type) {
-			case Breakpoint::BREAKPOINT: {
-				bpRow = createTableRow(BREAKPOINT);
-				setBreakpointStatus(BREAKPOINT, bpRow, true);
+QMap<QString, BreakpointData>::iterator
+BreakpointViewer::scanBreakpointData(const Breakpoint& bp)
+{
+	BreakpointData::Type type = bp.type == Breakpoint::BREAKPOINT ? BreakpointData::BREAKPOINT
+	                         : (bp.type == Breakpoint::CONDITION  ? BreakpointData::CONDITION : BreakpointData::WATCHPOINT);
+	if (map[type].isEmpty()) return map[type].end();
+
+	// assumed premise: a new bp will never repeat same name again
+	if (QMap<QString, BreakpointData>::iterator it; (it = map[type].find(bp.id)) != map[type].end() && it->index == -1) {
+		// if name remains the same, bp has not changed.
+		return it;
+	}
 
-				bool currentSlot = breakpoints->inCurrentSlot(bp);
-				QBrush color = palette().brush(runState | currentSlot ?
-					QPalette::ColorGroup::Normal : QPalette::ColorGroup::Disabled,
-					QPalette::WindowText);
+	return map[type].end();
+}
 
-				QTableWidgetItem* item1 = bpTableWidget->item(bpRow, ENABLED);
-				item1->setCheckState(Qt::Checked);
+void BreakpointViewer::connectBreakpoint(const QString& id, BreakpointData& data)
+{
+	map[data.type].insert(id, data);
+}
 
-				QTableWidgetItem* item2 = bpTableWidget->item(bpRow, BP_ADDRESS);
-				item2->setForeground(color);
-				item2->setText(hexValue(bp.address, 4));
+void BreakpointViewer::sync()
+{
+	// don't reload if generated its own update.
+	if (selfUpdating) {
+		selfUpdating = false;
+		return;
+	}
 
-				QTableWidgetItem* item3 = bpTableWidget->item(bpRow, T_CONDITION);
-				item3->setForeground(color);
-				item3->setText(bp.condition.trimmed());
+	bool empty[3] = {false, false, false};
 
-				QTableWidgetItem* item4 = bpTableWidget->item(bpRow, SLOT);
-				item4->setForeground(color);
-				item4->setText(QString("%1/%2")
-					.arg(bp.ps == -1 ? QChar('X') : QChar('0' + bp.ps))
-					.arg(bp.ss == -1 ? QChar('X') : QChar('0' + bp.ss)));
+	for (auto type : {BreakpointData::BREAKPOINT, BreakpointData::WATCHPOINT, BreakpointData::CONDITION}) {
```

Comment:
This could work:
```
    for (int num = 0; num < BreakpointData::ALL; ++num) {
        BreakpointData::Type type = static_cast<BreakpointData::Type>(num);
```
because using `Type` directly in the for-loop would require two static_casts:
```
    for (BreakpointData::Type type = BreakpointData:BREAKPOINT; type < BreakpointData::ALL;
            type = static_cast<BreakpointData::Type>(static_cast<int>(type) + 1)) {
```
Too verbose and not very clear.